stm32/usb: Add support for using TinyUSB stack.

andrewleech opened the pull request:
### Summary

This PR adapts the stm32 port to allow switching from STM USB stack to TinyUSB stack. 

Using TinyUSB improves consistancy with other micropython ports and brings in the ability to use the runtime usb definition support recently added to other TinyUSB based ports.

TinyUSB can be enabled in a build with `#define MICROPY_HW_TINYUSB_STACK (1)` 
It's also enabled by default with `MICROPY_PREVIEW_VERSION_2`

From the command line it can be enabled with:
``` bash
make -C ports/stm32 BOARD=PYBV11 CFLAGS_EXTRA='-DMICROPY_HW_TINYUSB_STACK=1'
```

### Testing

Development and testing so far has been on the STM32WB55 and STM32F765

I've been more recently using and testing this in conjunction with [shared/tinyusb: Add support for USB Network (NCM) interface. #16459](https://github.com/micropython/micropython/pull/16459) on the STM32F765 where the builtin CDC and NCM endpoints are both working very well.

I haven't yet done significant testing of dynamic / runtime usb devices.

### Trade-offs and Alternatives

Currently the TinyUSB integration matches other ports for features, so CDC is enabled for repl use and [machine.USBDevice](https://docs.micropython.org/en/latest/library/machine.USBDevice.html) support is available by default.

This means however the existing stm32 / pyb usb support has been removed, including [pyb.USB_VCP](https://docs.micropython.org/en/latest/library/pyb.USB_VCP.html) and [pyb.usb_mode](https://docs.micropython.org/en/latest/library/pyb.html#pyb.usb_mode). 

Features like mass storage, hid and a second vcp could in theory be added via `machine.USBDevice` however this hasn't been tested and is a breaking change for the stm32 port.
 
TODO: test lightsleep behaviour / compared with https://github.com/micropython/micropython/pull/8304#issuecomment-2303453613

dpgeorge commented:
> For now I (intend to) maintain back-compatibility with the `MICROPY_HW_ENABLE_STM_USB` flag to revert to the stm stack instead of tinyusb.

Yes, please do allow support for either the STM or TinyUSB stack.

andrewleech commented on shared/runtime/pyexec.c:
```diff
@@ -583,7 +583,7 @@ int pyexec_friendly_repl(void) {
+        #if MICROPY_HW_ENABLE_USB && 0
         if (usb_vcp_is_enabled()) {
             // If the user gets to here and interrupts are disabled then
             // they'll never see the prompt, traceback etc. The USB REPL needs
```
@dpgeorge This block in the shared pyexec code appears to be stm32 specific! Other ports use the `MICROPY_HW_ENABLE_USBDEV` define and don't define `usb_vcp_is_enabled` etc. 
This little feature to re-enable any disabled interrupts once a device hits repl does sound convenient, would it be worth extending this to all ports? Or does this feel like something users haven't needed and should be removed?

dpgeorge replied:
I reckon we can just remove this.  It's pretty hacky and doesn't generalise well to other ports.

pi-anl committed:
shared/runtime/pyexec: Remove legacy USB IRQ enable code.

It's very STM32 USB stack specific and doesn't generalise well to other
ports.  So remove it.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

andrewleech commented on ports/stm32/usbd_conf.c:
```diff
@@ -51,213 +51,230 @@ PCD_HandleTypeDef pcd_hs_handle;
-  * @retval None
-  */
-void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd) {
+void pyb_usbd_init_fs(void) {
```
make these static

dpgeorge commented:
Code size change for stm32 is still very high....

andrewleech commented:
> Code size change for stm32 is still very high....

Yes it is! I've had to do a fairly deep dive into the code / build to figure this one out, it's more than the tinyusb files being always compiled (to avoid promoting this config to a Makefile setting) - it seems most of this size difference is coming from a reordering in the linker/map causing changes in padding between symbols, or in some cases its' from which module first includes a common element etc. A minor example of this; This diff alone causes a +8 bytes in total flash use: 
<img width="923" height="460" alt="image" src="https://github.com/user-attachments/assets/a4c8e6ee-a1b0-490a-9348-d51fe82abef0" />

Regardless, it looks like I'll be able to fix this with reording some elements and adding more `#ifdef` around stuff I missed in the last update (eg. the new `static inline void mp_hal_wake_main_task_from_isr(void)` function in mphalport.h)

dpgeorge commented on ports/stm32/mpconfigport.h:
```diff
@@ -179,6 +179,9 @@
 #define MICROPY_FATFS_RPATH            (2)
 #define MICROPY_FATFS_MULTI_PARTITION  (1)
+#if MICROPY_HW_TINYUSB_STACK && CFG_TUD_MSC
+#define MICROPY_FATFS_MAX_SS           (4096)
```
Ah, yes, that makes sense!

andrewleech commented:
I believe the size issue (when building with stm usb stack) has been resolved finally. Though I want to check exactly how it's now smaller...

On a positive note, switching to TinyUSB actually reduces the size by ~5.4KB.

dpgeorge commented:
> I believe the size issue (when building with stm usb stack) has been resolved finally. Though I want to check exactly how it's now smaller...

Yes, that's a bit concerning that it's smaller.  Probably something has been unintentionally removed/disabled.

dpgeorge commented:
I tried to test this on PYBD_SF6, PYBV10 and NUCLEO_WB55 but it did not work:
- PYBD_SF6 uses both FS and HS ports, and the main one is the HS (secondary port), and from what I can tell this isn't supported by TinyUSB (or at least our configuration).  So I gave up on this board.
- PYBV10 I expected to work because it's simple and just uses FS.  But enabling TinyUSB, nothing happens, no USB device appears.
- NUCLEO_WB55 I also expected to work, but similar to PYBV10 no USB device appears.  The UART REPL is still there and working, so the board starts up, but no native USB.

To enable TinyUSB I changed `mpconfigboard_common.h`:
```diff
--- a/ports/stm32/mpconfigboard_common.h
+++ b/ports/stm32/mpconfigboard_common.h
@@ -142,7 +142,7 @@
 
 // Select whether TinyUSB or legacy STM stack is used to provide USB.
 #ifndef MICROPY_HW_TINYUSB_STACK
-#define MICROPY_HW_TINYUSB_STACK (0)
+#define MICROPY_HW_TINYUSB_STACK (1)
 #endif
```
and it's definitely using that config option.

Did I do something wrong?

dpgeorge commented on ports/stm32/mpconfigboard_common.h:
```diff
@@ -256,6 +274,8 @@
 // Windows needs a different PID to distinguish different device configurations.
 #ifndef MICROPY_HW_USB_VID
 #define MICROPY_HW_USB_VID              (0xf055)
+#define MICROPY_HW_USB_PID              (0x9802)
```
Maybe add a comment that this is only used in the TinyUSB configuration (and eventually we'll need to change this value depending on whether MSC is enabled etc).

andrewleech commented:
@dpgeorge it turns out the WB55 / tinyusb issue was both more complicated and simpler than expected. 

The 16bit register access issue I found with tinyusb/1cf8e34ae5f had actually already been solved in a newer version of TinyUSB (I should have looked there first) but that alone didn't fix the issue.

Claude then figured out that the wrong USB base register was being used; in tinyusb it uses `#define FSDEV_REG_BASE USB_BASE` but for WB55 it needed to be `#define FSDEV_REG_BASE USB1_BASE` and that fixed micropython support.
... except then a recursive grep for `USB1_BASE` also found: 
```
./hw/mcu/st/cmsis_device_wb/Include/stm32wb55xx.h                  
969:#define USB1_BASE             (APB1PERIPH_BASE + 0x00006800UL)  // ((0x40000000UL) + 0x00006800UL)
1091:#define USB                 ((USB_TypeDef *) USB1_BASE)       
13070:#define USB_BASE                             (USB1_BASE)     
```
Turns out that's a newer version of the WB hal that tinyusb pulls, our one in microypthon is older and doesn't have https://github.com/STMicroelectronics/cmsis-device-wb/commit/d1b860584dfe24d40d455ae624ed14600dfa93c9#diff-a7f3443ed971e94fc67ce2c7fcdfde0030f573aeaaab4412180d4ef6d7ca7689R12165 
In our copy we have `#define USB_BASE (0x40005C00UL)` whereas `USB1_BASE` expands to `0x40006800UL` . 

TLDR no tinyusb PR is needed, but we need to update both tinyusb to the latest release _and_ [lib/stm32lib/CMSIS/STM32WBxx](https://github.com/micropython/stm32lib/pull/26)

dpgeorge commented:
Interesting!

Does that fix PYBV10 as well?

andrewleech commented:
I'm not sure yet, I actually don't have one at home (really should buy one) but I do have a F411 which I should be able to test. I obviosly need to clean up the build failures first regardless!

andrewleech commented:
> Maybe it's worth splitting out the TinyUSB update to a separate PR? Then it can be tested against existing TinyUSB-based ports.

I've reordered the branch to split this out into https://github.com/micropython/micropython/pull/18160

pi-anl committed:
stm32/usb: Add VBUS sensing configuration for TinyUSB on F4/F7.

TinyUSB's `dwc2_phy_init()` only sets the PWRDWN bit in the GCCFG register
but doesn't configure VBUS sensing, which is required for the DWC2 USB
controller to detect host connection on STM32F4 and STM32F7 series.

Without VBUS sensing configured, USB devices fail to enumerate on these
platforms when using the TinyUSB stack, while the legacy STM32 USB stack
works because it includes this configuration.

This commit adds VBUS sensing configuration in `pyb_usbd_init()` for
TinyUSB mode on STM32F4/F7:
- When VBUS detect pin is configured (typically PA9): Enable "B device"
  VBUS sensing (USB_OTG_GCCFG_VBUSBSEN)
- When no VBUS pin: Force VBUS valid (USB_OTG_GCCFG_NOVBUSSENS)

Tested on:
- NUCLEO-F429ZI: USB now enumerates successfully (CDC + MSC)
- NUCLEO-H563ZI: No regression (STM32H5 uses different register layout)

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

dpgeorge commented on ports/stm32/main.c:
```diff
@@ -88,6 +87,15 @@
+#endif
+
+#if MICROPY_HW_ENABLE_USB
+#include "usb.h"
```
Couldn't we just unconditionally include this particular header, like it was before this PR?  Keeps it simple.

dpgeorge commented on ports/stm32/mphalport.c:
```diff
@@ -53,6 +67,13 @@ MP_WEAK int mp_hal_stdin_rx_chr(void) {
         if (dupterm_c >= 0) {
             return dupterm_c;
         }
+        #if MICROPY_HW_ENABLE_USBDEV && MICROPY_HW_USB_CDC && MICROPY_HW_TINYUSB_STACK
```
Can remove `MICROPY_HW_ENABLE_USBDEV &&`.

dpgeorge commented on ports/stm32/usbd_conf.c:
```diff
@@ -56,18 +62,17 @@ PCD_HandleTypeDef pcd_hs_handle;
+{
     #if MICROPY_HW_USB_FS
-    if (hpcd->Instance == USB_OTG_FS) {
+    #if !MICROPY_HW_TINYUSB_STACK
```
Suggest `#if MICROPY_HW_STM_USB_STACK` instead.

dpgeorge commented on ports/stm32/usbd_conf.c:
```diff
@@ -290,6 +316,8 @@ void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd) {
     #endif // MICROPY_HW_USB_HS
 }
 
+#if !MICROPY_HW_TINYUSB_STACK
```
Suggest `#if MICROPY_HW_STM_USB_STACK` instead.

pi-anl committed:
stm32/usb: Add support for using TinyUSB stack.

This commit adapts the stm32 port to allow switching from STM USB stack to
TinyUSB stack.

Using TinyUSB improves consistancy with other MicroPython ports and brings
in the ability to use the runtime USB definition support recently added to
other TinyUSB based ports.

By default the existing STM USB stack is used.  TinyUSB can be enabled in a
board configuration with:

    #define MICROPY_HW_TINYUSB_STACK (1)

Or, it can be enabled from the command line with:

    make -C ports/stm32 CFLAGS_EXTRA='-DMICROPY_HW_TINYUSB_STACK=1'

Signed-off-by: Andrew Leech <andrew@alelec.net>

pi-anl committed:
stm32/usb: Add TinyUSB Mass Storage support.

Implements USB MSC functionality for STM32 port when using TinyUSB stack,
supporting both internal Flash and SD card storage mediums.

Signed-off-by: Andrew Leech <andrew@alelec.net>

dpgeorge commented:
The CI is failing.  I suggest the following patch to hopefully fix that:
```diff
--- a/ports/stm32/mpconfigboard_common.h
+++ b/ports/stm32/mpconfigboard_common.h
@@ -635,8 +635,13 @@
 #define MICROPY_HW_BDEV_WRITEBLOCKS(src, bl, n) spi_bdev_writeblocks(MICROPY_HW_BDEV_SPIFLASH, (src), MICROPY_HW_BDEV_SPIFLASH_OFFSET_BLOCKS + (bl), (n))
 #endif
 
+// Define the FATFS maximum sector size.
+#ifndef MICROPY_FATFS_MAX_SS
 #if defined(STM32N6)
 #define MICROPY_FATFS_MAX_SS                    (4096)
+#else
+#define MICROPY_FATFS_MAX_SS                    (512)
+#endif
 #endif
```

dpgeorge commented:
Thanks for updating, this is looking pretty good now.  I'll do some final testing on my side, then hopefully merge.

dpgeorge commented on ports/stm32/modos.c:
```diff
@@ -52,7 +52,7 @@ bool mp_os_dupterm_is_builtin_stream(mp_const_obj_t stream) {
            #endif
-           #if MICROPY_HW_ENABLE_USB
+           #if MICROPY_HW_STM_USB_STACK
            || type == &pyb_usb_vcp_type
```
OK!  A follow-up PR sounds good for that.

dpgeorge reviewed (approved):
Tested on PYBV11, with STM USB and TinyUSB.  It works.

iabdalkader commented:
@dpgeorge Might be unrelated, but there's an issue with `mp_usbd_cdc.c:mp_usbd_cdc_tx_strn`: if the FIFO is full *and* the device is disconnected, it will never timeout or break. Admittedly, I can only reproduce this with our firmware, probably because I send FIFO-sized packets with our protocol stuff, but suppose upstream can get into that state somehow, you need to run the `tud_task` for `tud_cdc_connected()` to change, to even run that inner timeout loop.

Removing `if (tud_cdc_connected())` completely, or just adding the following to the loop tail, fixes this issue for me:

```C
        if (tud_task_event_ready()) {
            mp_usbd_task();
        }
```

Is this something you're interested in improving/fixing, can I send a PR for this?

andrewleech commented:
@iabdalkader thanks for looking into that, it's certainly something I'm interested in having improved! These kind of corner-cases are the hardest to nail down and you've done most of the hard work already.

I don't remember the full context of that area of code clearly but a lot of the tud_cdc_connected() logic there was to support the fifo being able to be filled and maintained before it's connected so that on first connected you can get the startup banner etc displayed that was captured earlier in startup. It was supposed to properly fifo that buffer though to only show the most recent data.

I expect that removing if (tud_cdc_connected()) will break the startup buffering, however adding the tud_task_event_ready() check feels like relatively low overhead to add to fix the lock up!
